### PR TITLE
fix(ecs): filter out non-active ECS services

### PR DESCRIPTION
**Pull request by iamhopaul123**
Applies similar fix as https://github.com/aws/copilot-cli/pull/5062 to ecs service.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/aws/ecs/service.go [hunk 1]
```diff
@@ -123,32 +123,66 @@ type ServiceArn string
...
+
+func (s *ServiceArn) parse() (*parsedServiceArn, error) {
 	serviceArn := string(*s)
 	parsedArn, err := arn.Parse(serviceArn)
 	if err != nil {
-		return "", err
+		return nil, err
+	}
+	if parsedArn.Service != EndpointsID {
+		return nil, fmt.Errorf("expect an ECS arn: got %q", serviceArn)
```
Comment (dannyrandall) [inline internal/pkg/aws/ecs/service.go]:
```suggestion
		return nil, fmt.Errorf("expected an ECS arn, but got %q", serviceArn)
```

File: internal/pkg/aws/ecs/service.go [hunk 2]
```diff
@@ -123,32 +123,66 @@ type ServiceArn string
...
 // will return my-project-test-Cluster-9F7Y0RLP60R7
 func (s *ServiceArn) ClusterName() (string, error) {
+	parsedArn, err := s.parse()
+	if err != nil {
+		return "", err
+	}
+	return parsedArn.clusterName, nil
+}
+
+type parsedServiceArn struct {
```
Comment (dannyrandall) [inline internal/pkg/aws/ecs/service.go]: i wonder if it makes sense to do something like:
```
type ServiceARN struct {
  // AccountID, Partition, Region, etc
}

func ParseServiceARN() (ServiceARN, error) {
  // contents of parse()
}
```

then users of the struct can pull fields without `parse()`ing multiple times

File: internal/pkg/ecs/ecs.go [hunk 3]
```diff
@@ -449,10 +450,23 @@ func (c Client) serviceARN(app, env, svc string) (*ecs.ServiceArn, error) {
 	if len(services) == 0 {
 		return nil, fmt.Errorf("no ECS service found with tags %s", tags.String())
 	}
-	if len(services) > 1 {
+	arns := make([]string, len(services))
+	for i := range services {
+		arns[i] = services[i].ARN
+	}
+	serviceARN := ecs.ServiceArn(arns[0])
```
Comment (dannyrandall) [inline internal/pkg/ecs/ecs.go]: is this making the assumption that all of tagged the services are in the same cluster? isn't it possible that's not true?

Comment (iamhopaul123) [reply internal/pkg/ecs/ecs.go]: They are in the same cluster i think for this use case. We can make another function or modify this one if we need a different behavior in the other place

File: internal/pkg/aws/ecs/ecs.go [hunk 4]
```diff
@@ -333,13 +333,35 @@ func (e *ECS) ActiveClusters(arns ...string) ([]string, error) {
...
 	}
 
 	return active, nil
+}
 
+// ActiveServices returns the subset of service arns that have an ACTIVE status.
+func (e *ECS) ActiveServices(clusterARN string, serviceARNs ...string) ([]string, error) {
+	resp, err := e.client.DescribeServices(&ecs.DescribeServicesInput{
+		Cluster:  aws.String(clusterARN),
+		Services: aws.StringSlice(serviceARNs),
```
Comment (bvtujo) [inline internal/pkg/aws/ecs/ecs.go:360]: What happens if there are no arns specified? Is this an error, or does the DescribeServices call describe all services on the cluster?

Can we add a unit test looking for this behavior?

File: internal/pkg/ecs/ecs.go (hunk 3 above)
Comment (bvtujo) [reply internal/pkg/ecs/ecs.go]: I think having to specify an app+env provides enough security, right? I don't think Copilot will let you make two envs of the same name in the same app. Though what happens if someone creates an app in tools account A, then an env in account B, then an app in tools account C, and an env in account B? Maybe this is out of scope 😂

File: internal/pkg/ecs/ecs.go [hunk 5]
```diff
@@ -449,10 +450,23 @@ func (c Client) serviceARN(app, env, svc string) (*ecs.ServiceArn, error) {
...
+	clusterArn, err := serviceARN.ClusterArn()
+	if err != nil {
+		return nil, fmt.Errorf("get cluster arn: %w", err)
+	}
+	active, err := c.ecsClient.ActiveServices(clusterArn, arns...)
+	if err != nil {
+		return nil, fmt.Errorf("check if clusters are active: %w", err)
```
Comment (bvtujo) [inline internal/pkg/ecs/ecs.go]:
```suggestion
		return nil, fmt.Errorf("check if services are active: %w", err)
```

**Review by bvtujo — CHANGES_REQUESTED** [review]

File: internal/pkg/ecs/ecs.go (hunk 3 above)
Comment (iamhopaul123) [reply internal/pkg/ecs/ecs.go]: Actually i'll change the logic

**Comment by iamhopaul123** [comment]
> I think having to specify an app+env provides enough security, right? I don't think Copilot will let you make two envs of the same name in the same app. Though what happens if someone creates an app in tools account A, then an env in account B, then an app in tools account C, and an env in account B? Maybe this is out of scope 

Good call i think i added some extra logic to catch this.

File: internal/pkg/aws/ecs/ecs.go (hunk 4 above)
Comment (iamhopaul123) [reply internal/pkg/aws/ecs/ecs.go:360]: very good call will add

File: internal/pkg/aws/ecs/ecs.go [hunk 6]
```diff
@@ -342,9 +345,23 @@ func (e *ECS) ActiveClusters(arns ...string) ([]string, error) {
 }
 
 // ActiveServices returns the subset of service arns that have an ACTIVE status.
```
Comment (bvtujo) [inline internal/pkg/aws/ecs/ecs.go:344]: Can we add a comment to suggest that this function should only be called on services in the same cluster?

**Review by bvtujo — APPROVED** [review]
lgtm just one tiny nit

File: internal/pkg/aws/ecs/ecs.go [hunk 7]
```diff
@@ -342,9 +345,24 @@ func (e *ECS) ActiveClusters(arns ...string) ([]string, error) {
...
-func (e *ECS) ActiveServices(clusterARN string, serviceARNs ...string) ([]string, error) {
+// Note that all services should be in the same cluster.
+func (e *ECS) ActiveServices(serviceARNs ...string) ([]string, error) {
+	if len(serviceARNs) == 0 {
```
Comment (dannyrandall) [inline internal/pkg/aws/ecs/ecs.go]: not sure if the api catches this error or just ignores it, but the max is 10

File: internal/pkg/aws/ecs/service.go [hunk 8]
```diff
@@ -163,26 +151,29 @@ func (s *ServiceArn) parse() (*parsedServiceArn, error) {
...
 // For example: arn:aws:ecs:us-west-2:1234567890:service/my-project-test-Cluster-9F7Y0RLP60R7/my-project-test-myService-JSOH5GYBFAIB
 // will return my-project-test-myService-JSOH5GYBFAIB
-func (s *ServiceArn) ServiceName() (string, error) {
-	parsedArn, err := s.parse()
-	if err != nil {
-		return "", err
-	}
-	return parsedArn.name, nil
+func (s *ServiceArn) ServiceName() string {
+	return s.name
```
Comment (dannyrandall) [inline internal/pkg/aws/ecs/service.go:169]: any reason to not just export this on the struct as `s.ServiceName`? (same for cluster name)

File: internal/pkg/aws/ecs/ecs.go (hunk 7 above)
Comment (dannyrandall) [reply internal/pkg/aws/ecs/ecs.go]: i kinda think - if the ECS api catches these errors (0 ARNs specified, > 10 specified), we shouldn't explicitly handle them here. at least for 0, it seems like we already have logic to catch that error in the caller.

if the ECS api has some weird behavior on them - like returning all services in the cluster or something - then we can handle them explicitly.

File: internal/pkg/aws/ecs/service.go (hunk 8 above)
Comment (iamhopaul123) [reply internal/pkg/aws/ecs/service.go:169]: Becuase everything in that struct should be private and init by the Parse function.

**Review by dannyrandall — APPROVED** [review]

File: internal/pkg/aws/ecs/ecs.go [hunk 9]
```diff
@@ -333,13 +333,47 @@ func (e *ECS) ActiveClusters(arns ...string) ([]string, error) {
...
+// Note that all services should be in the same cluster.
+func (e *ECS) ActiveServices(serviceARNs ...string) ([]string, error) {
+	var prevSvcArn *ServiceArn
+	for _, arn := range serviceARNs {
+		svcArn, err := ParseServiceArn(arn)
+		if err != nil {
+			return nil, err
+		}
+		if prevSvcArn != nil && prevSvcArn.clusterName != svcArn.clusterName {
+			return nil, fmt.Errorf("service %q and service %q should be in the same cluster", prevSvcArn.String(), svcArn.String())
```
Comment (joanniclaborde) [inline internal/pkg/aws/ecs/ecs.go:354]: Hi! I keep getting this error when doing `copilot service show`, and it's referencing old (deleted) clusters and services. Is it possible that this cluster check should be done after inactive services are filtered out?

**Comment by KollaAdithya** [comment]
Hey @joanniclaborde !

Which version of Copilot are you using. Fix for this is released in Copilot version `v1.29.1` 
https://github.com/aws/copilot-cli/releases/tag/v1.29.1

**Comment by joanniclaborde** [comment]
> Hey @joanniclaborde !
> 
> Which version of Copilot are you using. Fix for this is released in Copilot version `v1.29.1` https://github.com/aws/copilot-cli/releases/tag/v1.29.1

I'm using the latest version, v1.29.1.

**Comment by KollaAdithya** [comment]
Can you please paste the error message to help us debug the issue?

**Comment by joanniclaborde** [comment]
```
% copilot svc show                                                                                                                                                            
Service name: my-svc
✘ describe service my-svc: retrieve rollback alarm names: get service my-svc: check if services are active: service "arn:aws:ecs:us-east-1:1234567890:service/my-app-staging-Cluster-s63uAzrXXXXX/my-app-staging-my-svc-Service-c0iGyjmXXXXX" and service "arn:aws:ecs:us-east-1:1234567890:service/my-app-staging-Cluster-7tGl5VSXXXXX/my-app-staging-my-svc-Service-2DLOI9QXXXXX" should be in the same cluster
```

`my-app-staging-Cluster-s63uAzrXXXXX` is the current active cluster, while `my-app-staging-Cluster-7tGl5VSXXXXX` is an old deleted one.

I still get that same error message when I delete the whole app and recreate it (with the same names, of course).

Thanks!

**Comment by KollaAdithya** [comment]
Thanks for reporting the bug. I am able to reproduce it locally. We will work on the fix for this!

**Comment by dannyrandall** [comment]
Hey @joanniclaborde! We just released this fix [v1.30.0](https://github.com/aws/copilot-cli/releases/tag/v1.30.0)!🚀🎉 Thanks for reporting the issue!

**Pull request merged**